From Larry.Finger@lwfinger.net  Sun Dec  3 08:45:48 2006
Date: Sun, 03 Dec 2006 10:40:01 -0600
From: Larry Finger <Larry.Finger@lwfinger.net>
To: John Linville <linville@tuxdriver.com>, chrisw@sous-sol.org, stable@kernel.org
Cc: maxime@tralhalla.org, Michael Buesch <mb@bu3sch.de>, Stefano Brivio <st3@riseup.net>
Subject: softmac: fix unbalanced mutex_lock/unlock in ieee80211softmac_wx_set_mlme
Message-ID: <4572FDE1.mail33M13NDAZ@lwfinger.net>

From: Maxime Austruy <maxime@tralhalla.org>

Routine ieee80211softmac_wx_set_mlme has one return that fails
to release a mutex acquired at entry.

Signed-off-by: Maxime Austruy <maxime@tralhalla.org>
Signed-off-by: Larry Finger <Larry.Finger@lwfinger.net>
Signed-off-by: Chris Wright <chrisw@sous-sol.org>
---

John and Chris,

This error was introduced in the 2.6.19-rxX series and must be applied
to 2.6.19-stable and wireless-2.6.

Larry

 net/ieee80211/softmac/ieee80211softmac_wx.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- linux-2.6.19.orig/net/ieee80211/softmac/ieee80211softmac_wx.c
+++ linux-2.6.19/net/ieee80211/softmac/ieee80211softmac_wx.c
@@ -495,7 +495,8 @@ ieee80211softmac_wx_set_mlme(struct net_
 			printk(KERN_DEBUG PFX "wx_set_mlme: we should know the net here...\n");
 			goto out;
 		}
-		return ieee80211softmac_deauth_req(mac, net, reason);
+		err =  ieee80211softmac_deauth_req(mac, net, reason);
+		goto out;
 	case IW_MLME_DISASSOC:
 		ieee80211softmac_send_disassoc_req(mac, reason);
 		mac->associnfo.associated = 0;
